pytorch_lightning/trainer/trainer.py

```diff
@@ -852,11 +853,24 @@ def _load_checkpoint_weights(self):
         rank_zero_info(f"Loading model weights from checkpoint at {self._ckpt_path}")
         self.checkpoint_connector.restore_model_weights(self._ckpt_path)
 
+    def _maybe_switch_to_flexible_optimization_flow(self, model: "pl.LightningModule") -> None:
+        training_step_fx = getattr(model, "training_step")
+        if is_param_in_hook_signature(training_step_fx, "dataloader_iter", explicit=True):
```

Comment:
I would prefer to avoid activating on parameters signature. We should avoid as much as possible to rely on internal knowledge.

I would prefer Trainer(num_prefetch_batches=-1) to indicate the dataloader iterator should be passed directly.

Reply:
Will continue the discussion in https://github.com/PyTorchLightning/pytorch-lightning/issues/8867